### ISSUE TITLE
sly-edit-definition destroys windows

### ISSUE THREAD

**Issue by mfiano** [joaotavora/sly#121]
When using `M-.` to call `sly-edit-definition` and a sly-xref buffer is created in an existing window, to select a primary method for example, jumping to a primary method will close one of the windows.

Reproducible with `emacs -Q` and only loading sly package.

Steps to reproduce:

- `emacs -Q`
- split emacs into multiple windows with `C-x 2` and `C-x 3`. In this example I use a 2x2 grid, but any amount greater than 1 will do.
- load sly package
- run sly repl
- load a project via quicklisp etc
- in one window, open a file buffer for the loaded project which contains a callsite of a generic function having multiple primary methods.
- `M-.` on that callsite to have the sly-xref buffer displayed in another one of your emacs windows.
- select one of the primary methods lists in the sly-xref window, using `Enter` to accept.
- upon pressing enter you now have 1 less window. sly has destroy one in the process of destroying the xref buffer.

**Comment by joaotavora** [comment]
Confirmed. I can confirm the native emacs's own xref.el implementation does not do this. I am thinking of replacing SLY's implementation with emacs's own in the next SLY version, so I am wary of even touching this, because this is very "hairy" code.

However, I have prepared a commit in a separate branch that you could test...

**Comment by joaotavora** [comment]
The git branch is called 121-sly-xref-windows. Obviously,  I'm assuming a git installation of SLY, and also that you are comfortable switching branches in git. (nothing fancy...)

**Comment by mfiano** [comment]
Your fix is a lot better but it raises another issue. In the case of multiple choices (xref window appears), when you select a choice, the point stays in the previous window. The new xref target appears in the other window that sly-xref opened in, meanwhile, point is in original window `M-.` was issued in. You have to manually move window focus in order to continue to `M-.` / `M-,`.

**Comment by joaotavora** [comment]
I see, that's indeed a problem. 

I pushed two more commits to that branch. They both fix the latest issue, but the behaviour after selecting a target is different regarding the window they select the target in. 

I wonder if you can try them both and report separately about them here.

**Comment by mfiano** [comment]
Both commits seem to behave the same -- they both solve the issue. However, there is yet another issue. I am not sure if this coincides with SLIME or not, as I do not have experience with it, but, the xref choice you select is opened in the same window the xref buffer was created in. The original window `M-.` was issued in is not used. Additionally, since before any of your fixes, the window used for the xref buffer (and therefor the jumped to code buffer in these last 2 commits), seems to be completely random. If I have 9 window splits, it is completely random which one is used for the xref buffer.

However, the point does jump to the correct window/buffer upon selecting an xref choice now. I just find it unintuitive that a) the window is chosen randomly for the xref buffer, and b) the original window is not used for the target code jump.

**Comment by mfiano** [comment]
Assume I have a grid of 9 windows, labeled as such:

```
-------------
| A | B | C |
-------------
| D | E | F |
-------------
| G | H | I |
-------------
```

I am working in window 'A', and I issue `M-.` to jump to the definition of a function. Currently, this spawns a sly-xref buffer in a random window -- we'll say 'H':

```
-------------
| A |   |   |
-------------
|   |   |   |
-------------
|   | H |   |
-------------
```

Now, in the sly-xref window, I select the destination, and it jumps there. I am left with window 'A' where I started, and the point now at the target function located in the re-used window 'H'. This is a problem for 2 reasons.

- I issued `M-.` in window A, but now my point is in window 'H'. Issuing `M-,` to go back, and now I have a duplicate of the buffer located in windows 'A' and 'H'.

- Before I issued `M-.` originally, I had an important buffer in window 'H' that I expected to remain after sly-xref window is closed.

As you can see, the current functionality is rather unintuitive, and leaves me to manually show the original buffer that was in window 'H'. Ideally, sly should:

- Use a window location that is not so random for displaying the sly-xref buffer. Perhaps a window directly adjacent to the source window.

- After selecting a destination and closing sly-xref, the original buffer that was present in that window should be resumed. This means that the destination should be followed in the original window `M-.` was issued in.

**Comment by joaotavora** [comment]
I think you will find that window reuse strategies have been the only one
controversy well beyond Slime or SLY. See the documentation for
`display-buffer-alist` for a sense of the complexity involved.

 I believe that prior to your complaint (and indeed still in trunk) slime
and SLY behaved identically, but the randomness is still there.

This is why i will try to accommodate your legitimate objections in a
temporary commit, but then, as soon as possible, switch to emacs xref.el
implementation. SLY becomes simpler, code is reused, and more importantly,
you can complain to emacs :)

But meanwhile i believe you mistried the two most recent commits. (There
are three in total right now in the new branch).Try the butlast one, it's
less strange for a regular slime user than the last.

**Comment by mfiano** [comment]
Ok randomness aside, what about the other issue, about not reusing the source window for the jump?

**Comment by joaotavora** [comment]
This is what the butlast commit does! Please reset --hard your git head to
the commit starting with 0bda and try that.

**Comment by mfiano** [comment]
0bda was the first that I had tried.

**Comment by mfiano** [comment]
Well in this second set of tests. All together I tried 3 already.

**Comment by mfiano** [comment]
@joaotavora I created a screencast of the behavior I am experiencing on commit 0bda78.

First I test a destination that is located in the same file as the source. This just moves the point in the same buffer as expected. Then I try jumping to a destination located in another file, and it does not reuse the buffer as you say it does.

https://www.youtube.com/watch?v=A7Zr2Z5Ml3U

**Comment by joaotavora** [comment]
Ok, I'll analyse later at home. It might have to do with different emacs
versions. Thanks

**Comment by mfiano** [comment]
Sure thank you for your time :)

**Comment by joaotavora** [comment]
So I was trying out the latest git SLIME with

```
emacs -Q -l slime-autoloads.el -L . --eval '(setq inferior-lisp-program "sbcl")' -f slime
```

and it seems to be broken according to *both* of your criteria: (1) randomness of the choice for a reused xref window and (2) final selected window, but I'd like you to confirm that with you (hence the recipe line above) before we can confirm that this is not the direction we want to go.

On the other front, In 0bda786 I cannot confirm all of your reports: when I `M-.` on `cl:print-object` (easy test because it's a generic with lots of methods) in a multi-window scenario, the final window is different from the one where I `M-.`'ed *and* also different from the `*sly-xref for sbcl*` window. 

I tested this in emacs 25.1 and also the latest trunk.

Please confirm this bit.

However, I do confirm your report that `M-,` on that final window does not restore point to the window where you `M-.`'ed (assuming it is still there, it can very well not be). Would you like it to?

**Comment by mfiano** [comment]
> On the other front, In 0bda786 I cannot confirm all of your reports: when I M-. on cl:print-object (easy test because it's a generic with lots of methods) in a multi-window scenario, the final window is different from the one where I M-.'ed and also different from the *sly-xref for sbcl* window.

yes the final window is different from the one where I `M-.`, but no, it is the same as the sly-xref window. The original buffer that was in that window is not restored, and this is problematic for me.

**Comment by joaotavora** [comment]
Hi @mfiano, I just had a change to test on 25.2, and I strongly suspect you have a stale `sly.elc` file with a wrong definition of the internal elisp function `sly--pop-to-source-location`. Becuase `0bda786` really should do what you intent, if I understand you correctly.

To test against this, can you paste the definition at the end into the `*scratch*` buffer and evaluate it with `C-M-x` or `M-x eval-buffer RET`?

Then try again your problematic use case. FTR this is what I expect

1. start with two or more windows already open with x.lisp and z.lisp
2. find definition on x.lisp
3. xref pops up over window displaying z.lisp
4. choose some definition
5. x.lisp gets replaced by chosen target, say another position on x.lisp or some y.lisp
6. z.lisp is restored to the window where the xref buffer was.

```lisp
(defun sly--pop-to-source-location (source-location &optional method)
  (sly-move-to-source-location source-location)
  (let ((saved-point (point)))
    (cl-ecase method
      ((nil)     (switch-to-buffer (current-buffer)))
      (window    (pop-to-buffer (current-buffer) t))
      (delete-current (let ((original (selected-window))
                            (inhibit-redisplay t))
                        (pop-to-buffer (current-buffer) t)
                        (when (not (eq original (selected-window)))
                          (if (window-parameter original 'sly--xref-transient)
                              (delete-window original)
                            ;; for some reason `save-selected-window'
                            ;; is not exactly what we need here. See
                            ;; github issue #112.
                            ;;
                            (let ((saved (selected-window)))
                              (quit-window nil original)
                              (when (window-live-p saved)
                                (select-window saved)))))))
      (frame     (let ((pop-up-frames t)) (pop-to-buffer (current-buffer) t))))
    (sly--highlight-sexp)
    (goto-char saved-point)))
```

**Comment by mfiano** [comment]
I did exactly what you instructed, starting on 0bda78 and ensuring the function provided was defined just in case. I created 4 windows. I performed `M-.` in window 1. xref is displayed in window 3. i confirm a target in xref window, it is removed and a buffer is created in window 4. now i have the original buffer in window 1, and a new buffer in window 4 that has hijacked what was previously in window 4.

**Comment by mfiano** [comment]
I even tried this on 0bda78

```lisp
emacs -Q -l sly-autoloads.el -L . --eval '(setq inferior-lisp-program "sbcl")' -f sly
```

Same behavior

**Comment by joaotavora** [comment]
Ok this makes sense to me, regardless if it is what you actually want. But when I asked to to confirm this sentence:

>  the final window is different from the one where I M-.'ed and also different from the *sly-xref for sbcl* window.

you said:

> yes the final window is different from the one where I M-., but no, it is the same as the sly-xref window.

Which, if I gather correctly your very latest assessment, is not correct. 

Anyway, I understand what happes now on your part. Where would you like the target buffer to appear. In which of the four windows?

**Comment by joaotavora** [comment]
And also: please confirm that when the xref buffer disappears the window now displays the buffer that was there before.

**Comment by mfiano** [comment]
Oh sorry for the confusion. After speaking to some colleagues, we all agree that the target buffer should appear in the window I issued M-. in. Yes, xref disappears and restores old buffer.

**Comment by joaotavora** [comment]
OK, are you sure about this? Because Emacs's native xref behaves much like SLY does now in `0bda78`

If only one definition exists the target window is the one where M-. was issued, otherwise, a window is reused for the *xref* buffer and yet another window is reused for the target buffer.

**Comment by joaotavora** [comment]
I guess I can make SLY behave like you want, but this now becomes a feature and I have to decide what the default should be. I'm thinking of introducing a customization variable.

Anyway, can it be stated that 0bda786 is better than master, in your opinion?

The original problem that I reproduced (closing one of the windows) is now solved right? So I think I will push 0bda786 to master and work on your feature, unless you very strongly object (i.e. you think it is even worse than before...)

**Comment by mfiano** [comment]
Yes, I spoke to a few other people who use Sly, and my desired behavior makes sense to them. I am unaware of how Emacs handles xref, as I only use Sly's xref. If I issue `M-.` I expect the other buffers I have open in other windows to remain visible. I issued `M-.` somewhere, so I (and 2 other people) do not expect yet another window to be reused, thus hiding existing buffers I am working with.

**Comment by mfiano** [comment]
I would say that the original issue is solved, yes, thank you. It is still quite bothersome, but yes, this is a feature request at this point, so I don't have to constantly restore the previous buffer manually. Would you like me to create a new issue?

**Comment by joaotavora** [comment]
> Would you like me to create a new issue?

I guess, yes, if it's not too much trouble. Just paste the contents of your butlast comment as the desc.
